File: c++/QEasyWidgets/Windows/FramelessWindow/framelessWindow_linux.cpp
#include "framelessWindow.h"

#include <QApplication>
#include <QGuiApplication>
#include <QPainter>
#include <QStyle>
#include <QWindow>
#include <QScreen>
#include <QVBoxLayout>


/**
 * Linux-specific FramelessWindowBase implementation
 */

FramelessWindowBase::FramelessWindowBase(QWidget *parent, Qt::WindowFlags flags)
    : QWidget(parent, flags)
    , m_titleBar(nullptr)
    , m_mainLayout(nullptr)
    , m_maskWidget(nullptr)
    , m_windowState(Normal)
    , m_stretchable(true)
    , m_dragging(false)
    , m_edgeSize(3)
    , m_resizing(false)
    , m_resizeDirection(0) {
    setupUI();
}

void FramelessWindowBase::setupUI() {
    setWindowFlags(windowFlags() | Qt::FramelessWindowHint | Qt::WindowSystemMenuHint);

    m_mainLayout = new QVBoxLayout(this);
    m_mainLayout->setContentsMargins(0, 0, 0, 0);
    m_mainLayout->setSpacing(0);

    m_titleBar = new TitleBarBase(this);
    m_mainLayout->addWidget(m_titleBar);

    // Create mask widget for modal dialogs
    m_maskWidget = new QLabel(this);
    m_maskWidget->setStyleSheet("background-color: rgba(0, 0, 0, 111);");
    m_maskWidget->setAlignment(Qt::AlignCenter);
    m_maskWidget->hide();
}

void FramelessWindowBase::setMinimumSize(int width, int height) {
    QWidget::setMinimumSize(width, height);
}

void FramelessWindowBase::setFrameless(bool stretchable, bool dropShadowEffect) {
    m_stretchable = stretchable;
    setWindowFlags(windowFlags() | Qt::FramelessWindowHint);
    
    if (!stretchable) {
        m_edgeSize = 0;
    }
}

void FramelessWindowBase::setTitleBar(TitleBarBase *titleBar) {
    if (m_titleBar) {
        m_titleBar->deleteLater();
        m_titleBar->hide();
    }
    
    if (titleBar != nullptr) {
        m_titleBar = titleBar;
        m_titleBar->setParent(this);
        m_titleBar->show();
        m_titleBar->raise();
    } else {
        m_titleBar = nullptr;
    }
}

void FramelessWindowBase::showMask(bool show, const QString &maskContent) {
    if (show) {
        if (m_maskWidget) {
            m_maskWidget->raise();
            m_maskWidget->setText(maskContent);
            m_maskWidget->show();
            m_maskWidget->setGeometry(rect());
        }
    } else {
        if (m_maskWidget) {
            m_maskWidget->clear();
            m_maskWidget->hide();
        }
    }
}

void FramelessWindowBase::setWindowState(WindowState state) {
    if (m_windowState == state) return;

    if (state == Maximized) {
        maximizeWindow();
    } else if (state == Normal) {
        restoreWindow();
    }

    m_windowState = state;
}

bool FramelessWindowBase::checkIfDraggable(const QPoint &pos) const {
    return (0 < pos.x() && pos.x() < width() && 
            0 < pos.y() && pos.y() < m_titleBar->height()) 
            ? true : false;
}

void FramelessWindowBase::showEvent(QShowEvent *event) {
    QWidget::showEvent(event);
    emit showed();
}

void FramelessWindowBase::closeEvent(QCloseEvent *event) {
    QWidget::closeEvent(event);
    emit closed();
}

void FramelessWindowBase::resizeEvent(QResizeEvent *event) {
    QWidget::resizeEvent(event);
    emit rectChanged(rect());
    
    if (m_titleBar) {
        m_titleBar->resize(width(), m_titleBar->height());
    }
    
    if (m_maskWidget && m_maskWidget->isVisible()) {
        m_maskWidget->setGeometry(rect());
    }
    
    setCursor(Qt::ArrowCursor);
}

void FramelessWindowBase::moveEvent(QMoveEvent *event) {
    QWidget::moveEvent(event);
    emit rectChanged(rect());
}

void FramelessWindowBase::paintEvent(QPaintEvent *event) {
    QWidget::paintEvent(event);

    // Draw border if not maximized
    if (m_windowState != Maximized) {
        QPainter painter(this);
        painter.setPen(QColor(200, 200, 200));
        painter.drawRect(0, 0, width() - 1, height() - 1);
    }
}

void FramelessWindowBase::mousePressEvent(QMouseEvent *event) {
    if (event->button() == Qt::LeftButton && m_stretchable) {
        if (checkIfDraggable(event->pos())) {
            setCursor(Qt::OpenHandCursor);
            windowHandle()->startSystemMove();
            QApplication::instance()->postEvent(
                windowHandle(),
                new QMouseEvent(QEvent::MouseButtonRelease, QPoint(-1, -1), Qt::LeftButton, Qt::NoButton, Qt::NoModifier)
            );
        }
    }
    QWidget::mousePressEvent(event);
}

void FramelessWindowBase::mouseMoveEvent(QMouseEvent *event) {
    // Linux simplification: just update cursor
    QWidget::mouseMoveEvent(event);
}

void FramelessWindowBase::mouseReleaseEvent(QMouseEvent *event) {
    if (checkIfDraggable(event->pos())) {
        setCursor(Qt::ArrowCursor);
        windowHandle()->startSystemResize(Qt::Edges());
    }
    QWidget::mouseReleaseEvent(event);
}

void FramelessWindowBase::mouseDoubleClickEvent(QMouseEvent *event) {
    if (checkIfDraggable(event->pos()) && (event->buttons() & Qt::LeftButton)) {
        if (isMaximized()) {
            showNormal();
        } else {
            showMaximized();
        }
    }
    QWidget::mouseDoubleClickEvent(event);
}

bool FramelessWindowBase::nativeEvent(const QByteArray &eventType, void *message, qintptr *result) {
    return QWidget::nativeEvent(eventType, message, result);
}

void FramelessWindowBase::onMinimizeRequested() {
    showMinimized();
}

void FramelessWindowBase::onMaximizeRequested() {
    if (m_windowState == Maximized) {
        restoreWindow();
    } else {
        maximizeWindow();
    }
}

void FramelessWindowBase::onCloseRequested() {
    close();
}

void FramelessWindowBase::updateWindowGeometry() {
    if (m_windowState == Maximized) {
        QScreen *screen = QGuiApplication::primaryScreen();
        QRect screenGeometry = screen->availableGeometry();
        setGeometry(screenGeometry);
    }
}

void FramelessWindowBase::restoreWindow() {
    if (m_windowState == Maximized) {
        setGeometry(m_normalGeometry);
        m_windowState = Normal;
    }
}

void FramelessWindowBase::maximizeWindow() {
    if (m_windowState == Normal) {
        m_normalGeometry = geometry();
        updateWindowGeometry();
        m_windowState = Maximized;
    }
}

File: python/QEasyWidgets/Windows/FramelessWindow/framelessWindow_linux.py
from typing import Optional
from PySide6.QtCore import Qt, Signal, QPoint, QRect, QEvent
from PySide6.QtGui import QGuiApplication, QFont, QCursor, QMouseEvent, QShowEvent, QCloseEvent, QMoveEvent, QResizeEvent
from PySide6.QtWidgets import QApplication, QWidget, QLabel

from ...Common.Theme import BackgroundColorAnimationBase
from ...Common.StyleSheet import StyleSheetBase
from ...Components.Bar import TitleBarBase

##############################################################################################################################

class WindowBase(BackgroundColorAnimationBase):
    '''
    '''
    showed = Signal()
    closed = Signal()

    langChanged = Signal()

    rectChanged = Signal(QRect)

    edge_size = 3 # 窗体边缘尺寸（出现缩放标记的范围）

    def __init__(self,
        min_width = 630, # 窗体的最小宽度
        min_height = 420, # 窗体的最小高度
        *args, **kwargs
    ):
        super().__init__(*args, **kwargs)

        self.titleBar = TitleBarBase(self)

        self.mask = QLabel(self)
        self.rectChanged.connect(self.mask.setGeometry)
        self.mask.setStyleSheet('background-color: rgba(0, 0, 0, 111);')
        self.mask.setAlignment(Qt.AlignCenter)
        self.mask.setFont(QFont('Microsoft YaHei', int(min_height / 10), QFont.Bold))
        self.mask.hide()

        self.resize(min_width, min_height)

    def _check_ifdraggable(self, pos) -> bool:
        return (0 < pos.x() < self.width() and 0 < pos.y() < self.titleBar.height()) if self.titleBar is not None else False

    def _move_window(self, pos) -> None:
        self.windowHandle().startSystemMove()
        QApplication.instance().postEvent(
            self.windowHandle(),
            QMouseEvent(
                QEvent.MouseButtonRelease,
                QPoint(-1, -1),
                Qt.LeftButton,
                Qt.NoButton,
                Qt.NoModifier
            )
        )

    def _resize_window(self, pos, edges) -> None:
        self.windowHandle().startSystemResize(edges) if edges is not None else None

    def event(self, event: QEvent) -> bool:
        self.langChanged.emit() if event.type() == QEvent.LanguageChange else None
        return super().event(event)

    def mouseMoveEvent(self, event: QMouseEvent) -> None:
            return

    def mousePressEvent(self, event: QMouseEvent) -> None:
        if self._check_ifdraggable(event.position()) == True and event.buttons() == Qt.MouseButton.LeftButton:
            self.setCursor(Qt.CursorShape.OpenHandCursor)
            self._move_window(event.position())

    def mouseReleaseEvent(self, event: QMouseEvent) -> None:
        if self._check_ifdraggable(event.position()) == True:
            self.setCursor(Qt.CursorShape.ArrowCursor)
            self._resize_window(event.position(), None)

    def mouseDoubleClickEvent(self, event: QMouseEvent) -> None:
        if self._check_ifdraggable(event.position()) == True and event.buttons() == Qt.MouseButton.LeftButton:
            self.showNormal() if self.isMaximized() else self.showMaximized() #self.setWindowState(Qt.WindowState.WindowMaximized)

    def showEvent(self, event: QShowEvent) -> None:
        self.showed.emit()
        event.accept()

    def closeEvent(self, event: QCloseEvent) -> None:
        self.closed.emit()
        event.accept()

    def moveEvent(self, event: QMoveEvent) -> None:
        self.rectChanged.emit(self.rect())

    def resizeEvent(self, event: QResizeEvent) -> None:
        self.rectChanged.emit(self.rect())
        self.titleBar.resize(self.width(), self.titleBar.height()) if isinstance(self.titleBar, TitleBarBase) else None
        self.setCursor(Qt.CursorShape.ArrowCursor)

    def nativeEvent(self, eventType, message):
        if eventType.type() in (QEvent.MouseButtonPress, QEvent.MouseMove):
            border_width = self.edge_size
            left   = (eventType.globalPos() - self.pos()).x() < border_width
            top    = (eventType.globalPos() - self.pos()).y() < border_width
            right  = (eventType.globalPos() - self.pos()).x() > self.width() - border_width
            bottom = (eventType.globalPos() - self.pos()).y() > self.height() - border_width
            if True not in (left, top, right, bottom):
                pass
            elif eventType.type() == QEvent.MouseButtonPress and self.windowState() != Qt.WindowNoState:
                self._move_window(eventType.globalPos())
        return False

    def setFrameless(self, setStrechable: bool = True, setDropShadowEffect: bool = True) -> None:
        self.setWindowFlags(self.windowFlags() | Qt.FramelessWindowHint)

    def setTitleBar(self, titleBar: Optional[QWidget]) -> None:
        try:
            self.titleBar.deleteLater()
            self.titleBar.hide()
            StyleSheetBase.Bar.deregistrate(self.titleBar)
        except:
            pass
        if titleBar is not None:
            self.titleBar = titleBar
            self.titleBar.setParent(self) if self.titleBar.parent() is None else None
            self.titleBar.raise_() if self.titleBar.isHidden() else None
        else:
            self.titleBar = None

    def showMask(self, setVisible: bool, maskContent: Optional[str] = None) -> None:
        if setVisible:
            self.mask.raise_() if self.mask.isHidden() else None
            self.mask.setText(maskContent) if maskContent is not None else self.mask.clear()
            self.mask.show()
        else:
            self.mask.clear()
            self.mask.hide()

##############################################################################################################################